File: py/src/kiahk/coptic_date.py
"""CopticDate value type with validating constructor."""
from __future__ import annotations

from dataclasses import dataclass

from kiahk.errors import InvalidCopticDateError

# Coptic months 1..12 have 30 days. Month 13 (Nasie) has 5 days, or 6 in a
# leap year. Coptic leap year rule: Y mod 4 == 3 (Julian-style).


def _coptic_is_leap(c_year: int) -> bool:
    return c_year % 4 == 3


def _validate(year: int, month: int, day: int) -> None:
    if not 1 <= month <= 13:
        raise InvalidCopticDateError(f"coptic month must be 1..13, got {month}")
    if month <= 12:
        max_day = 30
    else:
        max_day = 6 if _coptic_is_leap(year) else 5
    if not 1 <= day <= max_day:
        raise InvalidCopticDateError(
            f"coptic day must be 1..{max_day} for year {year} month {month}, got {day}"
        )


@dataclass(frozen=True)
class CopticDate:
    year: int
    month: int
    day: int

    def __post_init__(self) -> None:
        _validate(self.year, self.month, self.day)

    def to_gregorian(self) -> "GregorianDate":  # noqa: F821
        from kiahk.algorithms import coptic_to_gregorian
        from kiahk.gregorian_date import GregorianDate

        return GregorianDate(*coptic_to_gregorian(self.year, self.month, self.day))

File: csharp/Kiahk/CopticDate.cs
namespace Kiahk;

/// <summary>A calendar date on the Coptic (Anno Martyrum) calendar.</summary>
/// <remarks>
/// Coptic months 1..12 are 30 days each. Month 13 (Nasie) is 5 days, or 6 in a
/// leap year (Y mod 4 == 3, Julian-style).
/// </remarks>
public sealed record CopticDate
{
    public int Year { get; }
    public int Month { get; }
    public int Day { get; }

    private static bool IsCopticLeap(int y) => y % 4 == 3;

    /// <summary>Construct and validate. Throws <see cref="InvalidCopticDateException"/> on bad input.</summary>
    public CopticDate(int year, int month, int day)
    {
        if (month < 1 || month > 13)
        {
            throw new InvalidCopticDateException(
                $"coptic month must be 1..13, got {month}");
        }
        int maxDay = 30;
        if (month == 13)
        {
            maxDay = IsCopticLeap(year) ? 6 : 5;
        }
        if (day < 1 || day > maxDay)
        {
            throw new InvalidCopticDateException(
                $"coptic day must be 1..{maxDay} for year {year} month {month}, got {day}");
        }
        Year = year;
        Month = month;
        Day = day;
    }

    /// <summary>Convert this Coptic date to a Gregorian date.</summary>
    public GregorianDate ToGregorian()
    {
        var r = Algorithms.CopticToGregorian(Year, Month, Day);
        return new GregorianDate(r.Year, r.Month, r.Day);
    }
}

// Partial-class extension: adds GregorianDate.ToCoptic() here so GregorianDate.cs
// doesn't have a dependency on CopticDate.
public sealed partial record GregorianDate
{
    /// <summary>Convert this Gregorian date to a Coptic date.</summary>
    public CopticDate ToCoptic()
    {
        var r = Algorithms.GregorianToCoptic(Year, Month, Day);
        return new CopticDate(r.Year, r.Month, r.Day);
    }
}
